libs/utils/android/gfxinfo.py

```diff
@@ -0,0 +1,91 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2017, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Parser for dumpsys gfxinfo output
+
+import ast, re
+
+def get_value(token):
+    try:
+        v = ast.literal_eval(token)
+    except:
+        return token
+    return v
+
+class GfxInfo(object):
+    """
+    Class for parsing and accessing GfxInfo output
+    """
+    __properties = {}
+
+    def __init__(self, filename):
+        """
+        Initialize gfxinfo parser object
+
+        filename: File containing output of gfxinfo
+        """
+        self.filename = filename
+        self.parse_gfxinfo()
+
+    def parse_gfxinfo(self):
+        """
+        Parser for gfxinfo output, creates a name/value pair
+        """
+        with open(self.filename) as f:
+            content = f.readlines()
+
+        for line in content:
+            line = line.rstrip()
+            if not ":" in line:
+                continue
+
+            tokens = line.split(":")
+            if (len(tokens[1]) < 1):
+                continue
+
+            tokens = [t.strip() for t in tokens]
+            tokens[0] = tokens[0].replace(" ", "_").lower()
+
+            if tokens[0] == 'janky_frames':
+                (frames, pc) = tokens[1].split(" ")
+                self.__properties["janky_frames"] = get_value(frames)
+                pc = re.sub("[\(\)\%]", "", pc)
+                self.__properties["janky_frames_pc"] = get_value(pc)
+                continue
+
+            self.__properties[tokens[0]] = get_value(tokens[1])
+
+    def __dir__(self):
+        """
+        List all available attributes including ones parsed from
+        gfxinfo output
+        """
+        return self.__properties.keys()
+
+    def __getattr__(self, name):
+        """
+        Get the gfxinfo property using the period operator
+        Ex: obj.number_missed_vsync
+        """
+        return self.__properties[name]
+
+    def __getitem__(self, name):
+        """
+        Get the gfxinfo property using the [] opertator
+        Useful for attributes like "50th_percentile" that can't
+        be fetched with the period operator
+        """
+        return self.__properties[name]
```

Comment:
I don't really like having both `__getitem__` and `__getattr__`, I would prefer just `__getitem__` (and also to add the `keys` method). This is personal taste though so let's see how @derkling feels.